distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Workbook.enso

```diff
@@ -289,12 +297,50 @@ type Excel_Workbook
 
     ## PRIVATE
     with_java_workbook : (Workbook -> Any) -> Any ! Illegal_State
-    with_java_workbook self f = Illegal_State.handle_java_exception <|
+    with_java_workbook self f = Illegal_State.handle_java_exception <| handle_invalid_location <| Empty_Sheet.handle_java_exception <|
         case self.excel_connection_resource_ref.get of
             Nothing ->
                 Error.throw (Illegal_State.Error "The workbook is being used after it was closed.")
             resource -> resource.with connection->
                 connection.withWorkbook f
 
+    ## GROUP Standard.Base.Input
+       ICON data_input
+       Reads one or more sheets into a Table.
+
+       Arguments:
+       - sheet_names: the names of the sheets to read. Defaults to all sheets.
+         An invalid `sheet_name` will be reported as a problem.
+       - headers: whether to use the first row as headers (default is
+         `Detect_Headers`) for each sheet.
+       - return: how to return the tables. Defaults to `Merged_Table` meaning
+         the tables are merged into a single table.
+       - on_problems: how to handle problems during reading. Defaults to
+         `Report_Warning`.
+    @sheet_names (s-> build_sheet_selector s)
+    read_many : Vector Text -> Headers -> Return_As -> Problem_Behavior -> Table
+    read_many self sheet_names:Vector=self.sheet_names (headers:Headers=..Detect_Headers) (return:Return_As=..Merged_Table) (on_problems:Problem_Behavior=..Report_Warning) =
```

Comment:
```suggestion
    read_many self sheet_names:Vector=self.sheet_names (headers:Headers=Headers.Detect_Headers) (return:Return_As=Return_As.Merged_Table) (on_problems:Problem_Behavior=..Report_Warning) =
```

I'm not sure, feel free to ignore, but I felt that in the same way as we do not `from ... import all` the things from the _current_ library, it feels like we may prefer to _not_ use autoscoping in regular library code, _especially_ as the types are present in the current scope. It removes some indirection of the autoscoping mechanism having to kick in.

Of course it is more than recommended to _do use_ autoscoping in tests.

But I guess it's also fine here, just adding my 2 cents.

Reply:
In these cases, I think it is good as it avoids repeating the type name. I think `(on_problems:Problem_Behavior=..Report_Warning)` is cleaner and we can remove the few constructors we export.

I agree on anything where any ambiguity. So I wouldn't do on a sum type I don't think.

Its a style thing so maybe one for us all to agree on.

Reply:
Okay.

My reasoning was that in libs code clarity > conciseness, so repeating the type name is no harm.

But it's probably good to use the style that is also suggested for end-users and for those autoscoping is most often preferred.

As for exporting - this does not matter - we have the `Problem_Behavior` available here as well, but that does not mean we have to export it from `Standard.Base`. End users will use autoscoping when changing the value. As far as I know the scope for default values is resolved statically, so it matters that `Problem_Behavior` is available in the scope in the lib (which it is and will be since it is on the type annotation), so this style choice in no way affects what we need to export.

But overall I'm 100% OK with preferring autoscoping also in the library code. 

I guess it may only matter in very hot code paths where it could potentially affect performance (not sure to what extent). But in top-level methods like this one it should not be harmful at all.

Reply:
Yeah I spoke with Jaroslav on that and agree we may need to optimise that at some point. The impact is pretty small and once resolved once it will then be cached (at least until the cache is exhausted).